pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -750,10 +764,24 @@ func (c *Reconciler) updateRunsStatusDirectly(pr *v1beta1.PipelineRun) error {
 
 type getTimeoutFunc func(ctx context.Context, pr *v1beta1.PipelineRun, rprt *resources.ResolvedPipelineRunTask, c clock.PassiveClock) *metav1.Duration
 
-func (c *Reconciler) createTaskRun(ctx context.Context, rprt *resources.ResolvedPipelineRunTask, pr *v1beta1.PipelineRun, storageBasePath string, getTimeoutFunc getTimeoutFunc) (*v1beta1.TaskRun, error) {
+func (c *Reconciler) createTaskRuns(ctx context.Context, rprt *resources.ResolvedPipelineRunTask, pr *v1beta1.PipelineRun, storageBasePath string, getTimeoutFunc getTimeoutFunc) ([]*v1beta1.TaskRun, error) {
+	var taskRuns []*v1beta1.TaskRun
+	matrixCombinations := matrix.FanOut(rprt.PipelineTask.Matrix).ToMap()
+	for i, taskRunName := range rprt.TaskRunNames {
+		params := matrixCombinations[strconv.Itoa(i)]
+		taskRun, err := c.createTaskRun(ctx, taskRunName, params, rprt, pr, storageBasePath, getTimeoutFunc)
```

Comment:
It seems like this creates each taskrun sequentially and synchronously.

1. would it make sense to create taskruns in goroutines?
2. If some taskruns are successfully created and some are not, will the PipelineRun's status correctly reflect that? If this isn't too complicated, could you add a test case for this scenario?

Reply:
we could explore creating parallel taskruns in goroutines, but i'd prefer that we do that everywhere as opposed to matrix only - see we don't use goroutines when creating taskruns from next schedulable pipelinetasks (we use for loops as currently implemented in matrix):

https://github.com/tektoncd/pipeline/blob/9d60d0adf41e74f78029a0eaab73140fa4f7206a/pkg/reconciler/pipelinerun/pipelinerun.go#L687-L712

if we want to change that behavior, I suggest we do that outside of the scope of this work, what do you think?

Reply:
re: `PipelineRun` status - we have not handled yet handled that - this change is focused on creating the `TaskRuns` (and it's already XL lol) - adding the `ChildReferences` for those `TaskRuns` will come right after this

Reply:
ah ok, it's helpful to know that we're already creating multiple taskruns in one reconcile loop, sounds like that's a good starting point for matrix then too.

re: status, my main concern was that we'd end up in an inconsistent state in the scenario I just described (some taskrun creations failing and some succeeding), and that the reconciler would attempt to create the taskruns again in the next reconcile loop. I'd be curious to know how we currently handle this when creating TaskRuns for all the Tasks that are ready to be scheduled. However, it sounds like since matrix isn't usable yet, that's ok for now? Maybe we revisit this concern when you add support for status updates?

Reply:
the same thing applies for parallel taskruns in general, outside of the matrix, so looking at the codeblock referenced above I believe there will be a failure when one of parallel taskruns is created (and others are not created) which causes the rest to complete execution before the pipelinerun exits - we can write tests to show this 

so if we want to change that behavior, we can look into it more broadly for all parallel executions of taskruns/runs

Reply:
SGTM!